File: lib/core/location/tubigon_boundary.dart
import 'dart:convert';

import 'package:flutter/services.dart';

/// Municipality boundary sourced from the Philippine GeoRisk/PSA municipal
/// boundary layer (PSGC 0701245000). Longitude is stored before latitude in
/// GeoJSON, but the public API consistently accepts latitude first.
class TubigonBoundary {
  TubigonBoundary._(this.polygons);

  static const assetPath = 'backend/resources/data/tubigon_boundary.geojson';
  static const minLatitude = 9.884021;
  static const maxLatitude = 10.072237;
  static const minLongitude = 123.881415;
  static const maxLongitude = 124.029692;

  static Future<TubigonBoundary>? _cached;

  /// GeoJSON MultiPolygon -> polygons -> rings -> coordinates.
  final List<List<List<TubigonCoordinate>>> polygons;

  Iterable<List<TubigonCoordinate>> get outerRings => polygons
      .where((polygon) => polygon.isNotEmpty)
      .map((polygon) => polygon.first);

  static Future<TubigonBoundary> load() => _cached ??= _load();

  static Future<TubigonBoundary> _load() async {
    final source = await rootBundle.loadString(assetPath);
    return fromJson(jsonDecode(source) as Map<String, dynamic>);
  }

  static TubigonBoundary fromJson(Map<String, dynamic> feature) {
    final geometry = feature['geometry'] as Map<String, dynamic>?;
    if (geometry?['type'] != 'MultiPolygon') {
      throw const FormatException(
          'Tubigon boundary must be a GeoJSON MultiPolygon.');
    }
    final rawPolygons = geometry!['coordinates'] as List<dynamic>?;
    if (rawPolygons == null || rawPolygons.isEmpty) {
      throw const FormatException('Tubigon boundary contains no polygons.');
    }
    final polygons = rawPolygons.map((rawPolygon) {
      return (rawPolygon as List<dynamic>).map((rawRing) {
        return (rawRing as List<dynamic>).map((rawPoint) {
          final point = rawPoint as List<dynamic>;
          return TubigonCoordinate(
            latitude: (point[1] as num).toDouble(),
            longitude: (point[0] as num).toDouble(),
          );
        }).toList(growable: false);
      }).toList(growable: false);
    }).toList(growable: false);
    return TubigonBoundary._(polygons);
  }

  bool contains({required double latitude, required double longitude}) {
    if (latitude < minLatitude ||
        latitude > maxLatitude ||
        longitude < minLongitude ||
        longitude > maxLongitude) {
      return false;
    }
    for (final polygon in polygons) {
      if (polygon.isEmpty ||
          !_ringContains(polygon.first, latitude, longitude)) {
        continue;
      }
      final insideHole = polygon
          .skip(1)
          .any((ring) => _ringContains(ring, latitude, longitude));
      if (!insideHole) return true;
    }
    return false;
  }

  bool _ringContains(
    List<TubigonCoordinate> ring,
    double latitude,
    double longitude,
  ) {
    var inside = false;
    for (var current = 0, previous = ring.length - 1;
        current < ring.length;
        previous = current++) {
      final a = ring[previous];
      final b = ring[current];
      if (_onSegment(a, b, latitude, longitude)) return true;
      final crosses = (a.latitude > latitude) != (b.latitude > latitude);
      if (crosses) {
        final intersection = (b.longitude - a.longitude) *
                (latitude - a.latitude) /
                (b.latitude - a.latitude) +
            a.longitude;
        if (longitude < intersection) inside = !inside;
      }
    }
    return inside;
  }

  bool _onSegment(
    TubigonCoordinate a,
    TubigonCoordinate b,
    double latitude,
    double longitude,
  ) {
    const epsilon = 1e-9;
    final cross = (longitude - a.longitude) * (b.latitude - a.latitude) -
        (latitude - a.latitude) * (b.longitude - a.longitude);
    if (cross.abs() > epsilon) return false;
    final minLongitude = a.longitude < b.longitude ? a.longitude : b.longitude;
    final maxLongitude = a.longitude > b.longitude ? a.longitude : b.longitude;
    final minLatitude = a.latitude < b.latitude ? a.latitude : b.latitude;
    final maxLatitude = a.latitude > b.latitude ? a.latitude : b.latitude;
    return longitude >= minLongitude - epsilon &&
        longitude <= maxLongitude + epsilon &&
        latitude >= minLatitude - epsilon &&
        latitude <= maxLatitude + epsilon;
  }
}

class TubigonCoordinate {
  const TubigonCoordinate({required this.latitude, required this.longitude});

  final double latitude;
  final double longitude;
}

File: backend/app/Support/TubigonBoundary.php
<?php

namespace App\Support;

use RuntimeException;

final class TubigonBoundary
{
    public const NAME = 'Municipality of Tubigon, Bohol, Philippines';

    public const PSGC = '0701245000';

    public const CENTER_LATITUDE = 9.9515287;

    public const CENTER_LONGITUDE = 123.9618897;

    private const MIN_LATITUDE = 9.884021;

    private const MAX_LATITUDE = 10.072237;

    private const MIN_LONGITUDE = 123.881415;

    private const MAX_LONGITUDE = 124.029692;

    // The authoritative land polygon ends at the shoreline. These tight
    // bounds cover only Tubigon's passenger-port service area so seaport pins
    // are not incorrectly rejected as outside the municipality.
    private const PORT_MIN_LATITUDE = 9.9545;

    private const PORT_MAX_LATITUDE = 9.9575;

    private const PORT_MIN_LONGITUDE = 123.9565;

    private const PORT_MAX_LONGITUDE = 123.9595;

    private ?array $polygons = null;

    public function contains(float $latitude, float $longitude): bool
    {
        if ($latitude < self::MIN_LATITUDE || $latitude > self::MAX_LATITUDE
            || $longitude < self::MIN_LONGITUDE || $longitude > self::MAX_LONGITUDE) {
            return false;
        }

        foreach ($this->polygons() as $polygon) {
            if (empty($polygon) || ! $this->ringContains($polygon[0], $latitude, $longitude)) {
                continue;
            }

            foreach (array_slice($polygon, 1) as $hole) {
                if ($this->ringContains($hole, $latitude, $longitude)) {
                    continue 2;
                }
            }

            return true;
        }

        return false;
    }

    public function containsPortServiceArea(float $latitude, float $longitude): bool
    {
        return $latitude >= self::PORT_MIN_LATITUDE
            && $latitude <= self::PORT_MAX_LATITUDE
            && $longitude >= self::PORT_MIN_LONGITUDE
            && $longitude <= self::PORT_MAX_LONGITUDE;
    }

    private function polygons(): array
    {
        if ($this->polygons !== null) {
            return $this->polygons;
        }

        $path = resource_path('data/tubigon_boundary.geojson');
        $feature = json_decode((string) file_get_contents($path), true);
        if (($feature['geometry']['type'] ?? null) !== 'MultiPolygon'
            || empty($feature['geometry']['coordinates'])) {
            throw new RuntimeException('The Tubigon municipal boundary asset is invalid.');
        }

        return $this->polygons = $feature['geometry']['coordinates'];
    }

    private function ringContains(array $ring, float $latitude, float $longitude): bool
    {
        $inside = false;
        $count = count($ring);
        for ($current = 0, $previous = $count - 1; $current < $count; $previous = $current++) {
            [$aLongitude, $aLatitude] = $ring[$previous];
            [$bLongitude, $bLatitude] = $ring[$current];

            if ($this->onSegment(
                (float) $aLatitude,
                (float) $aLongitude,
                (float) $bLatitude,
                (float) $bLongitude,
                $latitude,
                $longitude,
            )) {
                return true;
            }

            $crosses = ($aLatitude > $latitude) !== ($bLatitude > $latitude);
            if ($crosses) {
                $intersection = ($bLongitude - $aLongitude)
                    * ($latitude - $aLatitude)
                    / ($bLatitude - $aLatitude)
                    + $aLongitude;
                if ($longitude < $intersection) {
                    $inside = ! $inside;
                }
            }
        }

        return $inside;
    }

    private function onSegment(
        float $aLatitude,
        float $aLongitude,
        float $bLatitude,
        float $bLongitude,
        float $latitude,
        float $longitude,
    ): bool {
        $epsilon = 1e-9;
        $cross = ($longitude - $aLongitude) * ($bLatitude - $aLatitude)
            - ($latitude - $aLatitude) * ($bLongitude - $aLongitude);
        if (abs($cross) > $epsilon) {
            return false;
        }

        return $longitude >= min($aLongitude, $bLongitude) - $epsilon
            && $longitude <= max($aLongitude, $bLongitude) + $epsilon
            && $latitude >= min($aLatitude, $bLatitude) - $epsilon
            && $latitude <= max($aLatitude, $bLatitude) + $epsilon;
    }
}
